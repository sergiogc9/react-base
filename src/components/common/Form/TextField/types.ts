import { TextFieldProps } from '@sergiogc9/react-ui';

export type FormTextFieldProps = Omit<TextFieldProps, 'defaultValue' | 'name' | 'onBlur' | 'onChange' | 'value'> &
	Required<Pick<TextFieldProps, 'name'>>;
