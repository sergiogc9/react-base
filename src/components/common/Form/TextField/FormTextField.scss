@import "@src/scss/_globals.scss";
@import "@src/scss/_mixins.scss";
@import "@src/scss/_variables.scss";

.react-base-form-text-field {
	@include textInput("darker");
	label {
		font-size: 1.2rem;
	}
	.md-text-field {
		font-size: 1.3rem;
	}

	&.with-border {
		.md-text-field {
			padding: 0.5rem;
			border: 1px solid #000;
			line-height: 2.5rem;
		}
		hr.md-divider {
			display: none;
		}
	}
}
