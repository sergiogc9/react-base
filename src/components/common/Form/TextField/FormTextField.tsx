import React from 'react';
import TextField from 'react-md/lib/TextFields';
import get from 'lodash/get';
import has from 'lodash/has';
import isURL from 'validator/lib/isURL';

import { withT, TProps } from '@src/lib/i18n';
import { TextFieldElement } from '@src/types/form';
import { formFieldBaseValidation } from '@src/components/common/Form/Form';
import { useForceFieldValue } from '@src/lib/hooks';

import './FormTextField.scss';

export type ComponentProps = TProps & {
	element: TextFieldElement,
	forceValue?: string,
	error: boolean,
	onChangeText: (elementId: string, text: string, isValid: boolean) => void
};

const __validateTextField = (element: TextFieldElement, newValue: string) => {
	if (!formFieldBaseValidation(element, newValue)) return false;
	if (!element.validations) return true;
	if (element.validations.maxLength && element.validations.maxLength < newValue.length) return false;
	if (element.validations.minLength && element.validations.minLength > newValue.length) return false;
	if (element.validations.url && !isURL(newValue)) return false;
	return true;
};

const FormTextField = (props: ComponentProps) => {
	const { t, element, forceValue, error, onChangeText } = props;

	const [value, setValue] = React.useState(element.defaultValue || "");
	useForceFieldValue(forceValue, null, element, setValue, onChangeText, __validateTextField);

	const __onChangeHandler = React.useCallback((text: React.ReactText) => setValue(text.toString()), []);
	const __onBlurHandler = React.useCallback(
		(ev: any) => onChangeText(element.id, ev.target.value.trim(), __validateTextField(element, ev.target.value.trim())),
		[element, onChangeText]
	);

	return <TextField
		type="text"
		id={`formTextField-${element.id}`}
		className={`react-base-form-text-field ${element.className || ""}`}
		label={has(element, 'label.key') ? t(get(element, 'label.key')) : element.label}
		value={value}
		onChange={__onChangeHandler}
		onBlur={__onBlurHandler}
		required={!!element.required}
		error={error}
		disabled={element.disabled}
		{...element.inputProps}
	/>;
};

export default withT(React.memo(FormTextField));
