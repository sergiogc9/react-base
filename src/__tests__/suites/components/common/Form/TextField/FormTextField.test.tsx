import React from "react";
import { mount } from "enzyme";
import { act } from 'react-dom/test-utils';

import FormTextField from "@src/components/common/Form/TextField/FormTextField";
import { ComponentProps } from "@src/components/common/Form/TextField/FormTextField";
import TestHelper from "@src/__tests__/utils/Helper/TestHelper";

const element = TestHelper.textFieldElement;

let onChangeMock = jest.fn();
describe("Text field common component", () => {
	let wrapper: any;

	const updateComponent = (props: Partial<ComponentProps> = {}) => {
		act(() => {
			wrapper = mount(
				<FormTextField
					element={element}
					error={false}
					onChangeText={onChangeMock}
					{...props}
				/>);
		});
	};

	beforeEach(() => {
		onChangeMock = jest.fn();
		updateComponent();
	});

	it("Component default snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with label with key", () => {
		updateComponent({ element: { ...element, label: { key: "lang_key" } } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot after on change text input", () => {
		act(() => wrapper.find("TextField").at(0).props().onChange("newVal"));
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component on blur text input", () => {
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "newVal" } }));
		expect(onChangeMock).toHaveBeenCalledTimes(1);
	});

	it("Component on validate base validation without validation", () => {
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "newValue" } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "newValue", true);
	});

	it("Component on validate base validation with required but empty value", () => {
		updateComponent({ element: { ...element, required: true } });
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "" } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "", false);
	});

	it("Component on validate base validation", () => {
		updateComponent({ element: { ...element, validations: { fn: () => false } } });
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "newValue" } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "newValue", false);
	});

	it("Component on validate maxLength", () => {
		updateComponent({ element: { ...element, validations: { maxLength: 10 } } });
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "12345678901" } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "12345678901", false);
	});

	it("Component on validate minLength", () => {
		updateComponent({ element: { ...element, validations: { minLength: 5 } } });
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "123" } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "123", false);
	});

	it("Component on validate url", () => {
		updateComponent({ element: { ...element, validations: { url: true } } });
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "bad_url" } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "bad_url", false);
	});

	it("Component on validate correct", () => {
		updateComponent({ element: { ...element, validations: { minLength: 5 } } });
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: "12345" } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "12345", true);
	});

	it("Component snapshot after on force update", () => {
		wrapper.setProps({ forceValue: "forced!" });
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "forced!", true);
		expect(wrapper.html()).toMatchSnapshot();
	});
});
