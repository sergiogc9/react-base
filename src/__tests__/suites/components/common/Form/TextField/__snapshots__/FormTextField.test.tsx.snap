// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text field common component Component default snapshot 1`] = `"<div class=\\"md-text-field-container md-full-width md-text-field-container--input discover-form-text-field \\"><label for=\\"formTextField-textFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><input id=\\"formTextField-textFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div>"`;

exports[`Text field common component Component snapshot after on change text input 1`] = `"<div class=\\"md-text-field-container md-full-width md-text-field-container--input discover-form-text-field \\"><label for=\\"formTextField-textFieldId\\" class=\\"md-floating-label md-floating-label--floating md-text--secondary\\">Text label</label><input id=\\"formTextField-textFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"newVal\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div>"`;

exports[`Text field common component Component snapshot after on force update 1`] = `"<div class=\\"md-text-field-container md-full-width md-text-field-container--input discover-form-text-field \\"><label for=\\"formTextField-textFieldId\\" class=\\"md-floating-label md-floating-label--floating md-text--secondary\\">Text label</label><input id=\\"formTextField-textFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"forced!\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div>"`;

exports[`Text field common component Component snapshot with label with key 1`] = `"<div class=\\"md-text-field-container md-full-width md-text-field-container--input discover-form-text-field \\"><label for=\\"formTextField-textFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">lang_key</label><input id=\\"formTextField-textFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div>"`;
